postgres/attempt.go

```diff
@@ -540,6 +562,59 @@ func (w *worker) requestAttemptsForSpec(req coordinate.AttemptRequest, spec *wor
 	return attempts, err
 }
 
+// chooseMakeFailAttempts finds work units to do for a specific work
+// spec, honoring the work spec's MaxRetries field.  It doesn't make
+// sense to call this with maxRetries as zero; call
+// chooseAndMakeAttempts instead.
+func (w *worker) chooseMakeFailAttempts(
+	tx *sql.Tx,
+	spec *workSpec,
+	numUnits int,
+	maxRetries int,
+	now time.Time,
+	length time.Duration,
+) ([]coordinate.Attempt, error) {
+	var attempts []coordinate.Attempt
+	for len(attempts) < numUnits {
+		moreAttempts, err := w.chooseAndMakeAttempts(
+			tx, spec, numUnits-len(attempts), now, length)
+		if err != nil {
+			return nil, err
+		}
+		if len(moreAttempts) == 0 {
+			return attempts, nil
+		}
+
+		// For each of the (new) attempts, count the number of
+		// existing attempts for the work unit and maybe fail it.
+		// (It might be nice to do this in a batch, maybe moving
+		// this logic inside chooseAndMakeAttempts?)
+		for _, a := range moreAttempts {
+			// This should always be true but don't push it
```

Comment:
Because Go won't cast a `[]*attempt` to a `[]coordinate.Attempt`.  😢 Given typical values it's not actually a lot of wasted cycles to do the conversion so I'll go in and do that.